openassessment/workflow/workflow_batch_update_api.py

```diff
@@ -0,0 +1,600 @@
+
+    Returns:
+        list (PeerWorkflow): list of workfows not completed for > 7 days
+    """
+
+    filters = {
+        'created_at__lte': timezone.now() - datetime.timedelta(days=7),
+        'completed_at__isnull': True,
+    }
+    if course_id is not None:
+        filters['course_id'] = course_id
+    if item_id is not None:
+        filters['item_id'] = item_id
+    if submission_uuid is not None:
+        filters['submission_uuid'] = item_id
+
+    return PeerWorkflow.objects.filter(**filters)
+
+
+def get_workflow_update_data(peer_workflows):
+    """
+    Generates dictionary containing data required to update ORA workflows for all scopes.
+    This data structure is used as a local cache to avoid redundant DB queries during
+    batch workflow update process
+
+    Structure:
+    ```
+    {
+      "courses": [
+        {
+          "course_id": "value",
+          "course_settings": {},
+          "assessments": [
+            {
+              "item_id": "value",
+              "assessment_requirements": {},
+              "submissions": [
+                {
+                  "submission_uuid": "value"
+                }
+              ]
+            }
+          ]
+        }
+      ]
+    }
+    ```
+    """
+    workflow_update_data = {}
+    store = modulestore()
+    cache = set([])
+    for peer_workflow in peer_workflows:
+        try:
+            if peer_workflow.course_id not in cache:
+                # retrieve course block
+                course_block_key = UsageKey.from_string(peer_workflow.course_id)
+                course_block = store.get_item(course_block_key)
+
+                # retrieve openassessment block
+                ora_block_key = UsageKey.from_string(peer_workflow.item_id)
+                ora_block = store.get_item(ora_block_key)
+
+                if is_flexible_peer_grading_on(ora_block, course_block):
+                    # we are interested only in submissions for ORA with flexible peer grading configured
+                    course_settings = {
+                        'force_on_flexible_peer_openassessments': course_block.force_on_flexible_peer_openassessments}
+
+                    workflow_update_data = _add_course_data(
+                        workflow_update_data=workflow_update_data,
+                        course_id=peer_workflow.course_id,
+                        item_id=peer_workflow.item_id,
+                        submission_uuid=peer_workflow.submission_uuid,
+                        assessment_requirements=ora_block.workflow_requirements(),
+                        course_settings=course_settings)
+
+                cache.add(peer_workflow.course_id)
+                cache.add(peer_workflow.item_id)
+
+            elif peer_workflow.item_id not in cache:
+                # retrieve openassessment block
+                ora_block_key = UsageKey.from_string(peer_workflow.item_id)
+                ora_block = store.get_item(ora_block_key)
```

Comment:
I think this is also not correct. If an ora is in a course that is already in the cache, we fall into this block, but we don't check for whether flex peer grading is enabled for this item. You need to add a check for that here, otherwise you'll be adding info for all second/third/fourth/nth oras regardless of whether they should actually have been included

Reply:
Good findings, I have redesigned cache management approach a bit to make it more straightforward.